### PR TITLE
cslib_menu improvements

### PR DESCRIPTION
**Pull request by CareyJWilliams**

### CODE REVIEW

File: mygame.js [hunk 1]
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
...
+
+  if (this.stats.cslib_catch_choice) {
+    // automated choice selection for cslib_menu tests
+    if (this.stats.cslib_choice_select.length === 0)
+      throw new Error("Error: stuck in a choice menu but selections (cslib_choice_select) were exhausted.");
+
+    // Extract the next choice selection
+    var selections = this.stats.cslib_choice_select.split(",");
+    var thisSelection = parseInt(selections.shift());
+    this.stats.cslib_choice_select = selections.join(",");
```
Comment (diegoboston) [inline mygame.js:86]: interesting design - I guess it's the best way to interact with CS, even if a little clunky

File: mygame.js [hunk 2]
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
   } else {
     throw new Error(this.lineMsg() + message);
   }
-}
\ No newline at end of file
+}
+
+/*  AUTOMATIC CHOICE SELECTION #1
```
Comment (diegoboston) [inline mygame.js:30]: I am not sure I understand why you have #1 #2 #3, it seems like you have the code here (step #1) and the code needed in startup.txt (#2).

File: mygame.js [hunk 3]
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
...
+    method, which serves as the initial trigger to print the radio button
+    UI to the screen. We guard the old and new (automatic) functionality
+    on a global: cslib_catch_choice.
+
+    cslib_choice_select is another global that should contain the number
+    of the choice option to select (1 based), or a series of numbers,
+    separated by a comma, if your choices are nested.
+
+    A few helper variables are also created to help confirm that the length/number
+    (cslib_choice_count) of options and their text content (cslib_choice_n) is as expected.
```
Comment (diegoboston) [inline mygame.js:71]: didn't you just say this above?

File: mygame.js [hunk 4]
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
...
+
+    However, as randomtest.js defines its override _after_ mygame.js is
+    included, we have to override it via *script in startup.txt.
+
+    That means the code below is only the original source. The executed
+    code is found in startup.txt in a minimized form. If you update
+    the minimized code, please remember to update the source below as well.
+*/
+/*
+    Scene.prototype.choice = function choice(data) {
```
Comment (diegoboston) [inline mygame.js:122]: so, this must be copied to startup.txt? 

as I mentioned in the past, an alternative is to have EVERYTHING in startup.txt (even if hard to read, easier if we have the preprocessor)

File: scenes/startup.txt [hunk 5]
```diff
@@ -63,6 +84,10 @@
 
 *set implicit_control_flow true
 
+*if choice_randomtest
+	*comment override randomtest's Scene.prototype.choice so we can perform automated testing (original source in mygame.js)
+	*script Scene.prototype.choice=function(i){var t=["choice"];i&&(t=i.split(/ /));var n=this.parseOptions(this.indent,t),e=this;timeout=function(){println(""),e.renderOptions(t,n,function(i){e.standardResolution(i)})},this.finished=!0};
```
Comment (diegoboston) [inline scenes/startup.txt:100]: that's where you need the preprocessor :-)

**Review by danspinola — DISMISSED** [review]
As I far as I could tell, it looks awesome. 😁 But I'm not too familiar with JS, so I might have missed something.

File: mygame.js [hunk 6]
```diff
@@ -25,4 +25,111 @@ Scene.prototype.bug = function scene_bug(message) {
...
+  for (var i = 0; i < options.length; i++)
+    options[i].name = self.replaceVariables(options[i].name);
+
+  if (this.stats.cslib_catch_choice) {
+    // automated choice selection for cslib_menu tests
+    if (this.stats.cslib_choice_select.length === 0)
+      throw new Error("Error: stuck in a choice menu but selections (cslib_choice_select) were exhausted.");
+
+    // Extract the next choice selection
+    var selections = this.stats.cslib_choice_select.split(",");
```
Comment (diegoboston) [inline mygame.js:84]: do you need to support nested choices in your tests?

File: scenes/cslib_menu.txt [hunk 7]
```diff
@@ -105,38 +107,54 @@
...
+*return
+
+
+*comment ::: INTERNAL FUNCTIONS BELOW (DO NOT CALL DIRECTLY!) :::
+*label _build_simple_common
+*gosub _init_state
+*temp prefix "param"
+*temp filter ""
+*temp filter_value ""
+*temp page_count "@{(param_count <= const_items_per_page) 1|${(param_count / const_items_per_page)}}"
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: This stuff is getting really fancy - and it's really tricky for me to follow. Let's just test this stuff in and out.

File: scenes/cslib_menu.txt [hunk 8]
```diff
@@ -105,38 +107,54 @@
...
-*temp prefix "param"
-*temp filter ""
-*temp filter_value ""
 
-*label _init
-*comment CONSTANT CONFIG VALUES:
-*comment swap the next/prev buttons intelligently for smoother navigation
-*temp next_preference true
+*label _init_state
+*comment determines the next/prev option position for smoother navigation
```
Comment (diegoboston) [inline scenes/cslib_menu.txt:103]: Again, this is very ambitious. A version with no pagination would be easier to understand and debug. BTW, do you really expect a menu with multiple pages of choices?

File: scenes/cslib_menu.txt [hunk 9]
```diff
@@ -149,121 +167,202 @@
...
-		*goto _filter_loop
-*else
-	*if (cur_item < {prefix&"_count"})
-		*set cur_item + 1
-		*goto _filter_loop
-*if (item_count = 0) and not(allow_cancel)
-	*bug cslib_menu: menu construction failed as there were 0 matching items — consider allowing cancellation
+*temp total_item_count 0
+*temp cur_option_index 1
+*temp cur_item_index "@{const_invert_order ${{prefix&\"_count\"}}|1}"
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: Is this a multi-replace in a string? Wow, we are pushing the envelope here. This is actually a very powerful operator.

File: scenes/cslib_menu.txt [hunk 10]
```diff
@@ -149,121 +167,202 @@
...
+*temp items_off_page ((cur_page - 1) * const_items_per_page)
 
-*label _render
-*temp page_count 0
-*if (item_count < items_per_page)
-	*set page_count 1
+*label filter_loop
+*if (filter != "")
+	*gosub _apply_filter {((prefix&"_")&(filter&"_"))&cur_item_index} filter_value
+	*if not(cslib_ret)
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: does this mean cslib_ret != ""  ? I really lost track here :-(

File: scenes/cslib_menu.txt [hunk 11]
```diff
@@ -149,121 +167,202 @@
...
+*return
+
+*label _populate_page
+*comment populate the choice items and check the guards
+*temp page_item_count 0
+*temp slot 1
+*gosub _init_option_vars
+*label _populate_page_loop
+*if (slot <= const_items_per_page)
+	*if (prefix = "param")
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: Make sure you have tests both for arrays and params. Alternatively, we can choose to launch with a more limited (and better tested) set of functionality.

File: scenes/cslib_menu.txt [hunk 12]
```diff
@@ -149,121 +167,202 @@
...
+		*if (option_guard[slot])
+			*set page_item_count + 1
+			*set option_text[slot] "${param[offset]}"
+			*set option_ret_val[slot] offset
+	*else
+		*set option_guard[slot] ((option_slot_id[slot] > 0) and (const_items_per_page >= slot))
+		*if (option_guard[slot])
+			*set page_item_count + 1
+			*set option_text[slot] "${{(prefix&\"_\")&{\"option_slot_id_\"&slot}}}"
+			*set option_ret_val[slot] option_slot_id[slot]
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: this is REALLY hard to follow, CS doesn't help

File: scenes/cslib_menu.txt [hunk 13]
```diff
@@ -149,121 +167,202 @@
...
+	*if (allow_cancel) #${const_cancel_text}
 		*set cslib_ret 0
 		*goto _return
-*bug "_paginated_menu: unreachable"
+*bug "cslib_menu._render_choice: unreachable"
 *label _return
 *return
 
+
+*label _apply_filter
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: did you decide to keep the filters?

File: scenes/test_menu.txt [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+*gosub test_assert_equal 1 cslib_ret
+*gosub test_assert_equal "Hello" cslib_choice_1
+*gosub test_finish
+
+*gosub test_start "page-count invalid"
+*temp result_expected "Bug: cslib_menu._render_choice: this option should not be selectable"
+*set cslib_choice_select "1"
+*set cslib_catch_bug true
+*gosub_scene cslib_menu build_simple 1 2 3 4 5 6
+*set cslib_catch_bug false
```
Comment (diegoboston) [inline scenes/test_menu.txt:27]: do you want to assert on the message?

File: scenes/test_menu.txt [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+*gosub test_assert_equal "1" cslib_choice_3
+*gosub test_assert_equal "2" cslib_choice_4
+*gosub test_assert_equal "3" cslib_choice_5
+*gosub test_assert_equal "4" cslib_choice_6
+*gosub test_assert_equal "5" cslib_choice_7
+*gosub test_finish
+
+*gosub test_start "simple - multi-page (navigation)"
+*comment Select 'Next' then 'Next' then 'Previous', then '8' on Page 2
+*set cslib_choice_select "2,2,2,5"
```
Comment (diegoboston) [inline scenes/test_menu.txt:56]: oooh, you need multiple choices for pagination, right?

File: scenes/test_menu.txt [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+*gosub test_assert_equal 11 cslib_ret
+*gosub test_assert_equal "Page (3 / 3)" cslib_choice_1
+*gosub test_assert_equal "[i]<-- Previous[/i]" cslib_choice_2
+*gosub test_assert_equal "11" cslib_choice_3
+*gosub test_finish
+
+
+*comment :::::: ARRAY tests ::::::
+*comment array - init test data
+*set test_menu_opt_count 7
```
Comment (diegoboston) [inline scenes/test_menu.txt:84]: I won't stop this, but the filters are adding a layer of complexity - we don't have a strong design for 'objects' yet (for instance, I would feel this more natural: menu_opt[2]["visible"] than menu_opt["visible"][2])

File: scenes/test_menu.txt [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+*gosub test_assert_equal "6" cslib_choice_3
+*gosub test_assert_equal "7" cslib_choice_4
+*gosub test_assert_equal "8" cslib_choice_5
+*gosub test_assert_equal "9" cslib_choice_6
+*gosub test_assert_equal "[return]" cslib_choice_7
+*gosub test_finish
+
+*comment *gosub test_start "array - filter"
+*comment *set cslib_choice_select "2"
+*comment *gosub_scene cslib_menu build_array "test_menu_opt" false *comment "visible" "MATCH" true
```
Comment (diegoboston) [inline scenes/test_menu.txt]: are you missing a newline?

File: scenes/test_menu.txt [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+*comment *gosub test_assert_equal 3 cslib_choice_count
+*comment *gosub test_assert_equal 1 cslib_ret
+*comment *gosub test_assert_equal "Even" cslib_choice_1
```
Comment (diegoboston) [inline scenes/test_menu.txt]: are you not testing filters?

**Review by diegoboston — COMMENTED** [review]
This is very advanced, congratulations. Very nice trick to test the menu.
On the other side, this code is VERY hard to follow due to CS limitations --- great to have extensive tests.

I would consider launching with something simpler (without pagination or filters) but I won't stop you.   :-)

File: mygame.js (hunk 1 above)
Comment (CareyJWilliams) [reply mygame.js:86]: Feel free to suggest something less clunky!

File: scenes/test_menu.txt (hunk 14 above)
Comment (CareyJWilliams) [reply scenes/test_menu.txt:27]: Oops, yes. Good catch :)

File: scenes/test_menu.txt (hunk 15 above)
Comment (CareyJWilliams) [reply scenes/test_menu.txt:56]: Correct.

File: scenes/test_menu.txt (hunk 17 above)
Comment (CareyJWilliams) [reply scenes/test_menu.txt]: Huh, not sure what I've done there. Will take a look, thank you.

File: scenes/test_menu.txt (hunk 18 above)
Comment (CareyJWilliams) [reply scenes/test_menu.txt]: Apparently not. That's embarrassing, but probably what you get for pushing code at stupid o'clock.
Thanks for catching, I shall fix.

File: scenes/test_menu.txt (hunk 16 above)
Comment (CareyJWilliams) [reply scenes/test_menu.txt:84]: If we had an array module that allowed you to filter values _before_ passing them to cslib_menu, then I might agree, but as it stands there is no other way to present only a subset of options -- which is a crucial feature.

We even use it in library testing framework itself: https://github.com/ChoicescriptIDE/cslib/blob/9e69e00e65399c239ec5f9cb023fb7abc9b0b1ee/scenes/_test_entry.txt#L61 -- To only display test logs with content.

Removing this would make cslib_menu useless outside of situations where you knew every single option was selectable (which I imagine is actually the less frequent of use-cases).

--

The whole thing is meant to be a "beta" right? So let's see what people think? Better for people to say they don't like it and remove it, then never get feedback on it at all.

File: mygame.js (hunk 3 above)
Comment (CareyJWilliams) [reply mygame.js:71]: I can try to consolidate this a little more.

File: mygame.js (hunk 4 above)
Comment (CareyJWilliams) [reply mygame.js:122]: Hard to read (for users wanting to learn, not just contributors), near impossible to track (useful) changes through version control? Nope, sorry, I'm not sold. I'd sooner suggest forking ChoiceScript and making subtle changes, honestly.

File: mygame.js (hunk 6 above)
Comment (CareyJWilliams) [reply mygame.js:84]: Nested is probably the wrong word... 'subsequent' (though that would include nested).
But yes, this way I can check pagination works (both in terms of navigation, and that pages past page 1 are correct).
It's actually really quite fun...

File: scenes/cslib_menu.txt (hunk 7 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: It's basically a ternary operator: page_count = (param_count <= page_length) ? 1 : (param_count / page_length).
Happy to change variable names if that would help.

File: scenes/cslib_menu.txt (hunk 8 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt:103]: See my other comment somewhere above: I basically think that removing pagination would make this module redundant. Maybe you need to actually play with this patch yourself? Because I actually found it surprisingly easy to test pagination. It's quite fun actually.

You don't need to use the test framework, you can just fire up a normal choice, works a charm.

```
    *comment configure automatic choice selection
    *set cslib_catch_choice true
    *comment dictate which options (comma separated, for nested choices) should be selected
    *set cslib_choice_select "3"
    *fake_choice
      # Go east
        Eastward bound.
      # Go west
        Westward bound.
      # Go south
        Southward bound.
    *comment 'Southward bound.' will be printed here.
    *comment And helper variables will be populated:
    ${cslib_choice_count} == 3
    ${cslib_choice_1} == "Go east."
    ${cslib_choice_2} == "Go west."
    *comment etc.
    *comment and turn off automatic choice selection
    *set cslib_catch_choice false
```

File: scenes/cslib_menu.txt (hunk 9 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: Yeah, basically a ternary (again).

File: scenes/cslib_menu.txt (hunk 10 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: _apply_filter returns a boolean, so it's guarding on the result of that being false.

File: scenes/cslib_menu.txt (hunk 11 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: Test framework is here now, so feel free to suggest/write more tests. Should be very straight-forward now.

File: scenes/cslib_menu.txt (hunk 12 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: I'm open to suggestions...

File: scenes/cslib_menu.txt (hunk 13 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: Yes.

File: mygame.js (hunk 1 above)
Comment (diegoboston) [reply mygame.js:86]: I don't know if you can parse cslib_choice_select into a (JS) array and use that directly. Would that work as you pass it around? Or instead of splitting and rejoining, you may be able to do substring. Just ideas, not blockers.

File: mygame.js (hunk 4 above)
Comment (diegoboston) [reply mygame.js:122]: Agreed

File: scenes/cslib_array.txt [hunk 19]
```diff
@@ -0,0 +1,215 @@
+*bug cslib_array is a cslib module and must not be accessed directly
+
+*comment TODO: Decide array_attribute structure
```
Comment (diegoboston) [inline scenes/cslib_array.txt]: remove if not used

File: scenes/cslib_array.txt [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+*temp tar_max {p_array_ref_tar&"_max"}
+*set {p_array_ref_tar&"_count"} 0
+*temp src_index 1
+*temp tar_index 1
+*label _filter_by_routine_loop
+*if (src_index <= src_count)
+    *if (src_index > tar_max)
+        *bug cslib_array.filter_by_routine: Target array '${p_array_ref_tar}' of max size ${tar_max} is too small to hold all filtered values
+    *comment pass current element to filter function
+    *if (filter_by_routine)
```
Comment (diegoboston) [inline scenes/cslib_array.txt]: if we are not (yet) doing filter by value, I would simplify the code here - YAGNI   :-)

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

File: scenes/cslib_array.txt [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+*comment        p_array_ref_src(string): the name of the array to copy from
+*comment		p_array_ref_tar(string): the name of the array to copy to
+*comment	returns: 
+*comment		(number): the new _count value of the target array
+*label clone
+*params p_array_ref_src p_array_ref_tar
+*temp n 1
+*temp src_count {p_array_ref_src&"_count"}
+*temp tar_max {p_array_ref_tar&"_max"}
+*if (src_count > tar_max)
```
Comment (diegoboston) [inline scenes/cslib_array.txt:80]: great!

File: scenes/cslib_array.txt [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+*set {p_array_ref&"_count"} 0
+*return
+
+
+*comment SET
+*comment ------------------------------
+*comment Sets all elements in a given array to a given value,
+*comment and sets its _count to _max.
+*comment ------------------------------
+*label set_all
```
Comment (diegoboston) [inline scenes/cslib_array.txt:118]: this is very handy, I like it!

File: scenes/cslib_array.txt [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+*params p_array_ref p_value
+*temp end "@{param_count=3 ${param_3}|${{p_array_ref&\"_max\"}}}"
```
Comment (diegoboston) [inline scenes/cslib_array.txt]: VERY NICE!

File: scenes/cslib_array.txt [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+
+
+*comment PUSH
+*comment ------------------------------
+*label push
+*params p_array_ref p_value
+*gosub insert p_array_ref p_value ({{p_array_ref&"_count"}} + 1)
+*return
+
+*comment INSERT
```
Comment (diegoboston) [inline scenes/cslib_array.txt:162]: I was going to say: next we need some push/pop operation!

File: scenes/cslib_array.txt [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+*temp ret {p_array_ref&"_${p_index}"}
+*if (p_index = count)
+    *set {p_array_ref&"_count"} - 1
+*else
+    *gosub_scene cslib_array _shift_left p_array_ref p_index
+*set cslib_ret ret
+*return
+
+
+*comment SHIFT
```
Comment (diegoboston) [inline scenes/cslib_array.txt]: maybe add some more documentation on these operations

File: scenes/cslib_menu.txt (hunk 8 above)
Comment (diegoboston) [reply scenes/cslib_menu.txt:103]: thanks!

File: scenes/cslib_menu.txt [hunk 26]
```diff
@@ -105,165 +77,173 @@
...
+*label _populate_page
+*comment populate the choice items and check the guards
+*temp page_item_count 0
+*temp slot 1
+*gosub _init_option_vars
+*label _populate_page_loop
+*if (slot <= const_items_per_page)
+	*comment special handling for simple menus
+	*temp offset slot+(const_items_per_page * (cur_page -1))
+	*set option_guard[slot] ({prefix&"_count"} >= offset) and (const_items_per_page >= slot)
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: non actionable comment :-)

It's unfortunate we can't do {array_name}[idx], and that we have to bother with the underscore _ every time. It makes it unnecessarily hard to read

File: scenes/cslib_menu.txt [hunk 27]
```diff
@@ -105,165 +77,173 @@
...
+	*if (option_guard[slot])
+		*set page_item_count + 1
+		*set option_text[slot] "${{prefix&\"_${offset}\"}}"
```
Comment (diegoboston) [inline scenes/cslib_menu.txt]: would this work?
*set option_text[slot] {prefix & "_${offset}" }

also, if you change the prefix to include the _ (new *temp prefix_)  you can simply do

*set option_text[slot] {prefix_ & offset}

File: scenes/test_array.txt [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+*gosub test_assert_equal "JACKPOT" test_arr_1
+*gosub test_assert_equal "JACKPOT" test_arr_2
+*gosub test_assert_equal "JACKPOT" test_arr_3
+*gosub test_assert_equal "JACKPOT" test_arr_4
+*gosub test_assert_equal "JACKPOT" test_arr_5
+*gosub test_assert_equal "JACKPOT" test_arr_6
+*gosub test_assert_equal "JACKPOT" test_arr_7
+*gosub test_assert_equal "JACKPOT" test_arr_8
+*gosub test_assert_equal "JACKPOT" test_arr_9
+*gosub test_assert (((test_arr_max = test_arr_count)) and (test_arr_max = 9)) "test_arr_max (${test_arr_max}) and test_arr_count (${test_arr_count}) should both be equal to 9"
```
Comment (diegoboston) [inline scenes/test_array.txt:19]: consider using multiple assert (instead of AND), so you know which one failed

File: scenes/test_array.txt [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+*gosub_scene cslib_array remove "test_arr" 1
+*gosub test_assert_equal 0 test_arr_count
+*gosub test_finish
+
+
+*comment :::::: FILTER tests ::::::
+
+*comment our comparison routine applied to each array element to *comment determine whether it should be included in the filtered result
+*goto _skip_filter_test_is_even
+*label _filter_test_is_even
```
Comment (diegoboston) [inline scenes/test_array.txt:214]: NICE!

Comment (diegoboston) [inline scenes/test_array.txt:214]: NICE!

File: scenes/test_array.txt [hunk 30]
```diff
@@ -0,0 +1,243 @@
...
+*gosub test_start "filter by routine (too small)"
+*gosub_scene cslib_array set_all "test_arr" ""
+*comment pretend test_arr is smaller:
+*temp test_arr_max_store test_arr_max
+*set test_arr_max 3
+*set cslib_catch_bug true
+*gosub_scene cslib_array filter_by_routine "const_num" "test_arr" "test_array" "_filter_test_is_even"
+*set cslib_catch_bug false
+*temp test_arr_max test_arr_max_store
+*gosub test_finish "Bug: cslib_array.filter_by_routine: Target array 'test_arr' of max size 3 is too small to hold all filtered values" cslib_bug_message
```
Comment (diegoboston) [inline scenes/test_array.txt:242]: Very thorough tests, well done!

File: scenes/test_menu.txt (hunk 14 above)
Comment (diegoboston) [reply scenes/test_menu.txt:27]: I can't see the updated version with the fixed test

File: scenes/test_menu.txt (hunk 16 above)
Comment (diegoboston) [reply scenes/test_menu.txt:84]: Agreed

**Review by diegoboston — DISMISSED** [review]
This is really great. Please see what comments are actionable before merging.

File: scenes/test_array.txt (hunk 28 above)
Comment (CareyJWilliams) [reply scenes/test_array.txt:19]: Because it uses a custom assert message this actually makes it more obvious which one failed (the values are printed next to each), versus two generic "expected 9, received ?".

File: scenes/cslib_menu.txt (hunk 27 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: Thanks for this, it's a really good suggestion.
So I've added the _ to the prefix variable, and also moved the array count out into a variable. It's referenced a lot, and was indeed very ugly to look at.

Things look much better now:

```
*temp arr_count {{prefix&"count"}}
*temp page_count "@{(arr_count <= const_items_per_page) 1|${(arr_count / const_items_per_page)}}"
```

File: scenes/cslib_menu.txt (hunk 26 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt]: I've moved this out into an 'arr_count' temp, so it's a bit better now.

File: mygame.js (hunk 1 above)
Comment (CareyJWilliams) [reply mygame.js:86]: It would work, but it would produce more lines/verbosity, I think, as you'd need to check other that and cslib_choice_select weren't empty. I think I'll leave it as-is for now, but I'm happy to revisit.

File: .github/workflows/test-examples.yml [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+          path: cslib
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+          repository: 'dfabulich/choicescript'
+          path: choicescript
+      - name: Test Examples
+        run: |
+          cp cslib/examples/test-examples.sh .
+          ./test-examples.sh
```
Comment (diegoboston) [inline .github/workflows/test-examples.yml:31]: this is really neat!

**Review by diegoboston — APPROVED** [review]

File: scenes/cslib_menu.txt [hunk 32]
```diff
@@ -105,165 +77,174 @@
...
+*temp next_pref true
+*comment determines the direction in which options are displayed
+*temp const_invert_order false
+*comment the max amount of options to display per 'page'
+*temp const_items_per_page 5
+*if (const_items_per_page > 10)
+	*bug cslib_menu._init_state: const_items_per_page cannot be greater than 10
+*comment the text displayed on the cancel/return button
+*temp const_cancel_text "[return]"
+*temp arr_count {{prefix&"count"}}
```
Comment (diegoboston) [inline scenes/cslib_menu.txt:113]: do you need double braces?

**Review by diegoboston — APPROVED** [review]

File: scenes/cslib_menu.txt (hunk 32 above)
Comment (CareyJWilliams) [reply scenes/cslib_menu.txt:113]: No... Good spot. Although, if it's alright, I'll tidy that up in a separate commit. It's still functionally correct.

**Review by danspinola — APPROVED** [review]
